Add guide for enabling Shoot trust in Garden clusters with OIDC (#943)
gardener/documentation: website/documentation/guides/administer-shoots/garden-shoot-trust.md

---
title: Enabling Trust Between Shoot and Garden Clusters
description: "Set up OIDC-based trust so shoot service accounts can authenticate directly with the Garden cluster"
level: intermediate
index: 40
category: Security
scope: operator
---

## Overview

> This guide explains how to enable trust between the Garden cluster and a shoot cluster so that workloads running in the shoot can authenticate directly with the Garden API server using Kubernetes service account tokens (JWTs), removing the need for static credentials. Authorization decisions are then handled by RBAC in the Garden cluster. 

Reviewer: When using Gardener specific terminology, it's best to stick to the [Gardener glossary](https://gardener.cloud/docs/glossary/). As an example:
`Garden API server` -> should be `Gardener API server`
`Garden cluster` -> should be `garden cluster`

Could you please update the terms used throughout the topic?

> This guide explains how to enable trust between the garden cluster and a shoot cluster so that workloads running in the shoot can authenticate directly with the Gardener API server using Kubernetes service account tokens (JWTs), removing the need for static credentials. Authorization decisions are then handled by RBAC in the garden cluster. 

Reviewer 2: > Authorization decisions are then handled by RBAC in the garden cluster. 

Actually, it is not necessary the authorization to be handled by RBAC, it could be a webhook or some other plugin. I would change this sentence to be more inclusive/open towards other authorizers or remove it.

Example use cases:
- A shoot cluster that manages Gardener resources for a specific organization
- CI/CD pipelines running in a shoot that interact with the Gardener API

## How It Works
### Setup - TL;DR

The minimal shoot configuration requires two annotations:

```yaml
annotations:
    authentication.gardener.cloud/issuer: managed

>     authentication.gardener.cloud/trusted: 'true'

Reviewer 2: `authentication.gardener.cloud/trusted` requires the gardener installation to have oidc extension and the trust-configurator installed, right?

Author: Correct. I have briefly mentioned below that
> The following components help to enable this feature in the garden cluster:
> oidc-webhook-authenticator 
> garden-shoot-trust-configurator

Not sure if I should explicitly outline a sample gardener setup with garden extensions `shoot-oidc-service` and `garden-shoot-trust-configurator` and also the discovery server. What do you think?

Reviewer 2: Gardener operator deploys the discovery server unless it is disabled, i.e. when `Garden.Spec.VirtualCluster.Gardener.DiscoveryServer=nil`. I think you can add a `Prerequisite` subsection between `Overview` and `How It Works`, it will need a clarification that the prerequisites are fulfilled by the Garden admins, end-users are not expected to control enablement of the discovery server, oidc extension and shoot-trust configurator, so - they will have to check with their admin.

```

> Additionally, RBAC rules in the garden cluster should grant permissions to the shoot's service account identity which will perform requests to the garden cluster.

Reviewer 2: ```suggestion
Additionally, authorization in the garden cluster should be configured to grant permissions to the shoot's service account identity which will perform requests to the garden cluster.
```

### Authentication 

> The authentication flow, when a request is made is:

Reviewer: ```suggestion
When a request is made, the authentication flow is:
```

> ![Garden Shoot Trusted](./images/garden-shoot-trust.png)

Reviewer 2: If you can, remove the authorization part from the diagram.

- **[Managed Service Account Issuer (GEP-24)](https://github.com/gardener/enhancements/tree/main/geps/0024-shoot-oidc-issuer)**: Configures the shoot API server to sign service account tokens with a Gardener-managed key pair. The corresponding OIDC discovery documents and JWKS are served publicly by the Gardener Discovery Server.
- **[oidc-webhook-authenticator (OWA)](https://github.com/gardener/oidc-webhook-authenticator)**: A webhook token authenticator deployed in the garden cluster that validates OIDC tokens from trusted issuers.
- **[garden-shoot-trust-configurator](https://github.com/gardener/garden-shoot-trust-configurator)**: A controller that watches for shoots annotated as trusted and automatically creates the corresponding OIDC resources consumed by OWA.

## How to establish trust?

We will follow an example setup where a shoot will contain a health-reporter workload. It monitors the health status of all shoots in a Gardener project by querying the Garden API.

```bash
# Target shoot cluster
kubectl create namespace health-reporter
kubectl create serviceaccount health-reporter -n health-reporter
```

### Step 1: Enable Managed Issuer

> By enabling a managed service account issuer, Gardener configures the shoot API server to sign service account tokens with a managed key pair. The Gardener Discovery Server then exposes the corresponding OIDC discovery documents and JWKS at a publicly accessible URL, allowing the garden cluster to verify the token signatures.

Reviewer 2: > configures the shoot API server to sign service account tokens with a managed key pair

I think this is always the case, not only when the managed issuer is enabled, so maybe it can be removed (cc @dimityrmirchev for opinion)

Add the annotation to your shoot:

```yaml
annotations:
    authentication.gardener.cloud/issuer: managed
```

#### Verification
Verify in the shoot's status that the `service-account-issuer` is available:

```yaml
status:
  advertisedAddresses:
    - name: service-account-issuer
      url: >-

>         https://discovery.<service-account-issuer-hostname>/projects/<project-name>/shoots/<shoot-uid>/issuer

Reviewer 2: This advertised address is always set, by default it is based on the internal domain, and if I am not mistaken, users can set their own service account issuer, so the value can be controlled by themselves. If we have some doc that mentions the managed issuer URL, we can refer to it and here say to check the the shoot is using it.

```

We can find the publicly available discovery documents via the OIDC endpoint `{issuer-url}/.well-known/openid-configuration`:

```bash

> curl https://discovery.<domain>/projects/<project-name>/shoots/<shoot-uid>/issuer/.well-known/openid-configuration | jq .

Reviewer 2: If I am not mistaken, usually commands are prefixed with some prompt symbol to distinguish them from the command output. 
```suggestion
$ curl https://discovery.<domain>/projects/<project-name>/shoots/<shoot-uid>/issuer/.well-known/openid-configuration | jq .
```

{
  "issuer": "https://discovery.<domain>/projects/<project-name>/shoots/<shoot-uid>/issuer",
  "jwks_uri": "https://discovery.<domain>/projects/<project-name>/shoots/<shoot-uid>/issuer/jwks",
  "response_types_supported": [
    "id_token"
  ],
  "subject_types_supported": [
    "public"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ]
}
```

The service account tokens issued by the shoot's kube-API server will contain an `"iss"` claim that points to the configured issuer.

You can request a token with the `garden` audience (this must match the audience configured in the OIDC resource that the `garden-shoot-trust-configurator` creates).

```bash
kubectl create token health-reporter -n health-reporter --audience garden
```

Decode the JWT to verify its claims:

```json
{
  "aud": [
    "garden"
  ],

>   "iss": "https://discovery.<service-account-issuer-hostname>/projects/<project-name>/shoots/<shoot-uid>/issuer",

Reviewer 2: `discovery.<domain>` vs `discovery.<service-account-issuer-hostname>`, this will need to be unified accrross the document.

  "kubernetes.io": {
    "namespace": "health-reporter",
    "serviceaccount": {
      "name": "health-reporter",
      "uid": "<sa-uid>"
    }
  },
  "sub": "system:serviceaccount:health-reporter:health-reporter"
}
```

### Step 2: Annotate Shoot as Trusted

> To have the Garden cluster trust this shoot's issuer, add the trusted annotation:

Reviewer: ```suggestion
To have the Garden cluster trust this shoot's issuer, add the `trusted` annotation:
```

annotations:
    authentication.gardener.cloud/trusted: 'true'
```

The [`garden-shoot-trust-configurator`](https://github.com/gardener/garden-shoot-trust-configurator) watches for shoots with this annotation and automatically creates OIDC custom resources. These resources are consumed by the [`oidc-webhook-authenticator`](https://github.com/gardener/oidc-webhook-authenticator), which validates tokens presented to the Gardener API server.

In essence, this will create the following OIDC resource in the garden cluster:
```yaml
apiVersion: authentication.gardener.cloud/v1alpha1
kind: OpenIDConnect
metadata:
  labels:
    app.kubernetes.io/managed-by: garden-shoot-trust-configurator
  name: <project-ns>--<shoot-name>--<shoot-uid>
spec:
  audiences:
  - garden
  groupsClaim: groups
  groupsPrefix: 'ns:<project-ns>:shoot:<shoot-name>:<shoot-uid>:'
  issuerURL: "https://discovery.<service-account-issuer-hostname>/projects/<project-name>/shoots/<shoot-uid>/issuer"
  jwks:
    distributedClaims: true
  maxTokenExpirationSeconds: 7200
  supportedSigningAlgs:
  - RS256
  usernameClaim: sub
  usernamePrefix: 'ns:<project-ns>:shoot:<shoot-name>:<shoot-uid>:'
```

### Step 3: Configure RBAC

> Authentication (token verification) is now handled automatically, but the authenticated identity has no permissions by default. You must create RBAC rules in the garden cluster to authorize the requested actions.

Reviewer 2: ```suggestion
Authentication (token verification) is now handled automatically, but the authenticated identity has very restricted permissions by default, i.e. those available via the `system:authenticated` group. You must create RBAC rules, or configure any other authorizer that is in place, in the garden cluster to authorize the requested actions.
```

We need a `Role` that grants read access to `Shoot` resources and a `RoleBinding` that binds it to the shoot's ServiceAccount identity:

```bash
# Target garden cluster
kubectl apply -f - <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: shoot-health-reader
  namespace: garden-my-project
rules:
- apiGroups: ["core.gardener.cloud"]
  resources: ["shoots"]
  verbs: ["get", "list"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: health-reporter-binding
  namespace: garden-my-project
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: shoot-health-reader
subjects:
- kind: User
  name: "ns:<project-namespace>:shoot:<shoot-name>:<shoot-uid>:system:serviceaccount:health-reporter:health-reporter"
  apiGroup: rbac.authorization.k8s.io
EOF
```

> [!NOTE]
> The `garden-shoot-trust-configurator` configures a prefix to the token's `sub` and `groups` claim when deriving the name. This ensures identities from different shoots are always unique, even if they share the same ServiceAccount name. The resulting username has the form:
>
> ```yaml

> > ns:<project-namespace>:shoot:<shoot-name>:<shoot-uid>:<sub-claim>

Reviewer 2: For completeness, add the group prefix (even though it is available above).

> ```

### Step 4: Verification

Test that we are authenticated and authorized by requesting a token from the shoot and making a request to the Gardener API server:

```bash
# Target shoot cluster
TOKEN=$(kubectl create token health-reporter -n health-reporter --audience garden)

> curl -k -H "Authorization: Bearer $TOKEN" \

Reviewer 2: ```suggestion
curl -H "Authorization: Bearer $TOKEN" \
```
